src/CodeGen_PTX_Dev.cpp

```diff
@@ -791,6 +809,255 @@ bool CodeGen_PTX_Dev::supports_atomic_add(const Type &t) const {
+                                const bool load_b_ok = (load_b_var_1 && load_b_var_1->name == k_var_name) || (load_b_var_2 && load_b_var_2->name == k_var_name);
+
+                                if (load_a_ok && load_b_ok) {
+                                    // This matches A(k, y) * B(x, k) where both A and B
+                                    // are row-major
+
+                                    matmul_found = true;
+
+                                    A = load_a;
+                                    B = load_b;
+                                    C = matmul_load;
+
+                                    return;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        IRVisitor::visit(store);
+    }
+};
+
+ExtractTensorCoreOperations::ExtractTensorCoreOperations() {
+    thread_id_y = Call::make(Int(32), CodeGen_PTX_Dev::simt_intrinsic(".__thread_id_y"), std::vector<Expr>(), Call::Extern);
+    thread_id_x = Call::make(Int(32), CodeGen_PTX_Dev::simt_intrinsic(".__thread_id_x"), std::vector<Expr>(), Call::Extern);
+    block_id_y = Call::make(Int(32), CodeGen_PTX_Dev::simt_intrinsic(".__block_id_y"), std::vector<Expr>(), Call::Extern);
+    block_id_x = Call::make(Int(32), CodeGen_PTX_Dev::simt_intrinsic(".__block_id_x"), std::vector<Expr>(), Call::Extern);
+    block_dim_y = Call::make(Int(32), CodeGen_PTX_Dev::simt_intrinsic(".__block_dim_y"), std::vector<Expr>(), Call::Extern);
+    block_dim_x = Call::make(Int(32), CodeGen_PTX_Dev::simt_intrinsic(".__block_dim_x"), std::vector<Expr>(), Call::Extern);
+    block_size = block_dim_x * block_dim_y;
+}
+
+Stmt ExtractTensorCoreOperations::visit(const For *loop) {
+    const bool is_gpu_thread_var = CodeGen_GPU_Dev::is_gpu_thread_var(loop->name);
+    if (is_const_zero(loop->min) && is_const(loop->extent)) {
+        const int32_t loop_extent_value = loop->extent.as<IntImm>()->value;
+
+        if (is_gpu_thread_var) {
+            if (ends_with(loop->name, ".__thread_id_y")) {
+                wmma_M = loop_extent_value;
+            } else if (ends_with(loop->name, ".__thread_id_x")) {
+                wmma_N = loop_extent_value;
+            }
+        } else {
+            wmma_K = loop_extent_value;
+
+            // Shape m16n16k16
+            // TODO: Note this won't work unless the extent of the inner loop is constant
+            //       How can the shape be detected if the k is not constant, i.e, if k
+            //       is based on the dimentions of matrix A for example?
+            if (wmma_M == 16 && wmma_N == 16 && wmma_K % 16 == 0) {
+                const int32_t num_tiles_k = wmma_K / 16;
+                wmma_K = 16;
+
+                // Now check the loop body to confirm this is a matrix multiply expression
+                IsMatrixMultiply is_matrix_multiply{loop->name};
```

Comment:
I think maybe this should just be a function that digs through the body rather than a visitor? To make sure that this doesn't accidentally find a more deeply nested matrix multiply store? Or, what if the loop contained two separate matrix multiplications?

Reply:
This makes sense and the visitor is now converted into a a function in d2581cb46fb6e35c40be1d9aa640cdee436682a3